src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancies/AssetOccupancyManager.cs

```diff
@@ -0,0 +1,385 @@
+        }
+
+        return Task.FromResult(result);
+    }
+
+    protected virtual async Task<List<Schedule>> CalculateEffectiveSchedulesAsync(
+        DateTime startingDateTime,
+        DateTime endingDateTime,
+        List<AssetSchedule> appliedAssetSchedules,
+        Asset asset,
+        AssetCategory category)
+    {
+        var effectiveSchedules = new List<Schedule>();
+
+        foreach (var assetSchedule in appliedAssetSchedules
+                     .Where(x => x.PeriodUsable == PeriodUsable.Accept))
+        {
+            effectiveSchedules.Add(new Schedule
+            {
+                PeriodUsable = PeriodUsable.Accept,
+                StartingDateTime = assetSchedule.StartingDateTime <= startingDateTime
+                    ? startingDateTime
+                    : assetSchedule.StartingDateTime,
+                EndingDateTime = assetSchedule.EndingDateTime >= endingDateTime
+                    ? endingDateTime
+                    : assetSchedule.EndingDateTime,
+                TimeInAdvance = assetSchedule.TimeInAdvance
+            });
+        }
+
+        foreach (var assetSchedule in appliedAssetSchedules
+                     .Where(x => x.PeriodUsable == PeriodUsable.Reject))
+        {
+            // override accept policy
+            var rejectSchedule = new Schedule
+            {
+                PeriodUsable = PeriodUsable.Reject,
+                StartingDateTime = assetSchedule.StartingDateTime <= startingDateTime
+                    ? startingDateTime
+                    : assetSchedule.StartingDateTime,
+                EndingDateTime = assetSchedule.EndingDateTime >= endingDateTime
+                    ? endingDateTime
+                    : assetSchedule.EndingDateTime,
+                TimeInAdvance = assetSchedule.TimeInAdvance
+            };
+
+            var conflictedSchedule = effectiveSchedules.FirstOrDefault(x =>
+                x.IsTimeRangeIntersected(rejectSchedule.StartingDateTime, rejectSchedule.EndingDateTime));
+            if (conflictedSchedule is not null)
+            {
+                effectiveSchedules.Remove(conflictedSchedule);
+                // split conflicted schedule
+                effectiveSchedules.AddRange(
+                    conflictedSchedule.RemoveIntersectingTimeRange(
+                        rejectSchedule.StartingDateTime,
+                        rejectSchedule.EndingDateTime));
+            }
+
+            effectiveSchedules.Add(rejectSchedule);
+        }
+
+        effectiveSchedules.RemoveAll(x => x.IsEmptyTimeRange());
+
+        var effectiveTimeInAdvance = await CalculateEffectiveTimeInAdvanceAsync(asset, category);
+
+        foreach (var effectiveSchedule in effectiveSchedules)
+        {
+            effectiveSchedule.TimeInAdvance ??= effectiveTimeInAdvance;
+        }
+
+        // use default policy and it's fallback to fill the time gap 
+        var gaps = FindTimeGaps(startingDateTime, endingDateTime, effectiveSchedules);
```

Comment:
I'm confused. Could you please explain what this method and the blow code do? Privately communication is OK.